### PR TITLE
Filter os disk while creating/updating disk cr.

### PR DESCRIPTION
**Pull request by shovanmaity**
This PR contains plugable filter concept and os disk filter.
Filter interface is implemented in this PR.
In `Start()` of each filter we set range of check variables for that filter.
In `Include(*DiskInfo) bool` if property of disk matches with include value it returns true.
In `Exclude(*DiskInfo) bool` if property of disk does not match with exclude value it returns true.
If any of `Include()` or `Exclude()` returns true filter returns true.

**special notes for reviewer** -
I removed package level init for registering probes and filters. If I do like that then we will lost control of registering order as all of them are as go routine. And we want to register all filters then all probes so I have made separate register method of each probe and filters.

**Implementation steps** -
1. read `/proc/self/mounts` and find which partition is mounted on `/` path `ie - sda2`
2. read link `/sys/class/block/[partition-name]` ie - ```/sys/class/block/sda2```
this link looks like - 
```/sys/devices/pci0000:00/0000:00:1f.2/ata1/host0/target0:0:0/0:0:0:0/block/sda/sda2```
read `/` separated entry after ```block``` that is os disk name
3. we need to mount `/proc/self/mounts` inside container if we use self it points to currently running process so content of that file getting changed for that. We need to `mount /proc/1/mounts` this is the proc file of process 1 of host/node.
4. If process 3 is not able to get OS Disk uuid then it search for ```/etc/hosts``` mount point of containers ```/proc/self/mounts``` file. suppose ```/dev/sdc1``` on host is ```/var```, which is where ```/var/lib/docker``` resides and where Docker keeps certain data, such as the hosts file allocated to container. The hosts file is exposed as a bind mount inside the container at ```/etc/hosts``` mount point:
5. one new attribute IgnoreDiskUuid added in Controller struct at the time of event processing if the uuid of the disk is present in IgnoreDiskUuid then that event will not processed by event handler.
6. as a part of creating controller struct all ignored disk's uuid added in IgnoreDiskUuid field of Controller struct and cr creation of these disk is ignored.

For details - [Design doc](https://docs.google.com/document/d/1uyul5Sb6LL5PfXBE7tOUa4jSVLk-aPutZuP8kGRjGX0)

### CODE REVIEW

File: cmd/controller/osdisk.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (shovanmaity) [inline cmd/controller/osdisk.go]: @kmova This file contains functions to get uuid using mounts file and mount point. This is not related to controller. Thinking where to put this file inside util or some other package ... your suggestion

File: cmd/controller/filter.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		if filter.FilterState {
+			listFilter = append(listFilter, filter)
+		}
+	}
+	return listFilter
+}
+
+// ShouldProcess checks status for every registered filters if all filters allow
+// for their filter condition it returns true else it returns false
+func (c *Controller) ShouldProcess(diskDetails *DiskInfo) bool {
```
Comment (shovanmaity) [inline cmd/controller/filter.go]: ApplyFilter

File: cmd/controller/probe.go [hunk 3]
```diff
@@ -94,3 +94,13 @@ func (c *Controller) ListProbe() []*Probe {
 	}
 	return listProbe
 }
+
+// FillDetails lists registered probes and fills details from each probe
+func (c *Controller) FillDetails(diskDetails *DiskInfo) {
```
Comment (shovanmaity) [inline cmd/controller/probe.go]: FillDeviceDetails

File: cmd/filter/filter.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+const (
+	defaultEnabled  = true  // use in each filter to make it enable.
+	defaultDisabled = false // use in each filter to make it disable.
+)
+
+// RegisterdFilters contains register function of filters which we want to register
+var RegisterdFilters = []func(){osDiskFilterRegister}
+
+type registerFilter struct {
```
Comment (shovanmaity) [inline cmd/filter/filter.go]: remove prefix

File: cmd/controller/filter.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+// Start implements FilterInterface's Start()
+func (f *Filter) Start() {
+	f.Interface.Start()
+}
+
+// FilterInterface contains ApplyFilter interface and Start()
+type FilterInterface interface {
+	Start()
+	ApplyFilter
```
Comment (kmova) [inline cmd/controller/filter.go]: The struct ApplyFilter is clashing with the function ApplyFilter. Maybe the struct can be renamed to "Filters" or even totally eliminate it. bring up the Include and Exclude functions into the FilterInterace itself.

File: cmd/filter/osdiskfilter.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Include checks if disk uuid of passing DiskInfo struct is same as
+// uuid mention in osDiskFilter struct if it matches it returns true
+// else it returns false
+func (odf *osDiskFilter) Include(d *controller.DiskInfo) bool {
+	return odf.devPath == d.Path
+}
+
+// Include checks if disk uuid of passing DiskInfo struct is same as
```
Comment (kmova) [inline cmd/filter/osdiskfilter.go]: Fix these comments as they apply to OSdisk

File: cmd/probe/eventhandler.go [hunk 7]
```diff
@@ -43,21 +43,13 @@ func (pe *ProbeEvent) addDiskEvent(msg controller.EventMessage) {
...
-		for _, probe := range probes {
-			glog.Info("disk details filled by ", probe.ProbeName)
-			probe.FillDiskDetails(diskDetails)
-		}
-		diskApi := diskDetails.ToDisk()
-		oldDr := pe.Controller.GetExistingResource(diskList, diskDetails.ProbeIdentifiers.Uuid)
-		if oldDr != nil {
-			pe.Controller.UpdateDisk(diskApi, oldDr)
+		pe.Controller.FillDiskDetails(diskDetails)
+		if pe.Controller.ApplyFilter(diskDetails) {
```
Comment (kmova) [inline cmd/probe/eventhandler.go]: Before this call, add a comment on what is being one in apply filter

File: cmd/controller/filter.go (hunk 5 above)
Comment (shovanmaity) [reply cmd/controller/filter.go]: done.

File: cmd/filter/osdiskfilter.go (hunk 6 above)
Comment (shovanmaity) [reply cmd/filter/osdiskfilter.go]: done

File: cmd/probe/eventhandler.go (hunk 7 above)
Comment (shovanmaity) [reply cmd/probe/eventhandler.go]: done
